Src/IronPython/Runtime/Binding/MetaPythonFunction.cs

```diff
@@ -711,18 +709,42 @@ private class FunctionBinderHelper {
             /// a params list or the dictionary (or both).
             /// </summary>
             private Expression ValidateNotDuplicate(Expression value, string name, int position) {
-                EnsureParams();
+                List<Ast> statements = new List<Ast>();
 
-                return Ast.Block(
-                    Ast.Call(
-                        typeof(PythonOps).GetMethod(nameof(PythonOps.VerifyUnduplicatedByPosition)),
-                        AstUtils.Convert(GetFunctionParam(), typeof(PythonFunction)),    // function
-                        AstUtils.Constant(name, typeof(string)),                               // name
-                        AstUtils.Constant(position),                                           // position
-                        _paramsLen                                                        // params list length
-                        ),
-                    value
+                if (_userProvidedParams != null && position >= Signature.GetProvidedPositionalArgumentCount()) {
+                    EnsureParams();
+
+                    statements.Add(
+                        Ast.Call(
+                            typeof(PythonOps).GetMethod(nameof(PythonOps.VerifyUnduplicatedByPosition)),
+                            AstUtils.Convert(GetFunctionParam(), typeof(PythonFunction)),    // function
+                            AstUtils.Constant(name, typeof(string)),                         // name
+                            AstUtils.Constant(position),                                     // position
+                            _paramsLen                                                       // params list length
+                        )
                     );
+                }
+                if (_dict != null) {
+                    _needCodeTest = true;
+                    statements.Add(
+                        Ast.Call(
+                            typeof(PythonOps).GetMethod(nameof(PythonOps.VerifyUnduplicatedByName)),
+                            AstUtils.Convert(GetFunctionParam(), typeof(PythonFunction)), // function
+                            AstUtils.Constant(name, typeof(string)),                      // name
+                            AstUtils.Convert(_dict, typeof(PythonDictionary))             // params dict
+                        )
+                    );
+                }
+
+                // Return the value of the argument once we know that we don't have
+                // a duplicate
+                statements.Add(value);
```

Comment:
Maybe a little bit cleaner if you do an `if (statements.Count == 0) return value;` before this.